### PR TITLE
Improve integration test performance in scheduler

### PR DESCRIPTION
**Pull request by kerthcet**
#### What type of PR is this?
/kind feature
/sig scheduling

#### What this PR does / why we need it:

In part of the scheduler integration tests, for each test loop, we'll initialize the api server and scheduler, and tear them down after test finishes. Considering api server is quite time consuming to setup, I made some change:
1. Initialize api server for only once at the beginning of the test
2. Initialize scheduler every time in each test loop and tear it down after test

Take the `TestPostFilterPlugin` for example, before it costs ~ 53.273s, after, it costs ~ 14.079s.

- To achieve 1, I made some changes to the initialization util functions, see 4bec44c515a0a525ec7f31387aa4eb3af1ead1d8
- To achieve 2, some tests will setup both the apiserver and scheduler in the beginning, and call the reset() function manually after each test, together with deleting the pods created before, these are something I treat as micro-management, so 
    - I removed such logics, instead, I'll recreate the scheduler each time
    - Also removed the `reset()` functions, no need to care about the teardowns.
    - As a result, the test will slow for ~1s then, for we'll recreate the scheduler each time(only for those who create apiserver and scheduler at the beginning of the whole test)

Results:
```
before: ok k8s.io/kubernetes/test/integration/scheduler/plugins 320.507s
after: ok k8s.io/kubernetes/test/integration/scheduler/plugins 229.595s
``` 

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by kerthcet** [comment]
/hold
Something more complicated involved.

File: test/integration/util/util.go
Comment (kerthcet) [inline test/integration/util/util.go:459]: We add `eventBroadcaster.Shutdown()` both here and above because sometimes we'll not call `SchedulerCloseFn` in cleanup, then eventBroadcaster may not be closed properly.

**Comment by alculquicondor** [comment]
In some tests, we are removing the apiserver recreation for every case. In other tests, we are adding the scheduler recreation in every case.

Can you share what are the running times for the entire package before and after the change?

File: test/integration/scheduler/plugins/plugins_test.go
Comment (alculquicondor) [inline test/integration/scheduler/plugins/plugins_test.go:1451]: these look like unrelated refactorings, right?

File: test/integration/scheduler/plugins/plugins_test.go
Comment (alculquicondor) [inline test/integration/scheduler/plugins/plugins_test.go:2505]:
```suggestion
// cleanupTestSchedulerForFramework will close the scheduler and delete the nodes and pods.
```

Comment (alculquicondor) [reply test/integration/scheduler/plugins/plugins_test.go:2505]: although it sounds a bit odd. Maybe `cleanupSchedulerTestCase`?

**Comment by alculquicondor** [comment]
@pohly could you PTAL from sig-testing?

File: test/integration/util/util.go
Comment (kerthcet) [reply test/integration/util/util.go:190]: This is the same as other functions. I'd like to apply this to other functions as well, but in a followup.

**Comment by kerthcet** [comment]
> 

after: ok  	k8s.io/kubernetes/test/integration/scheduler/plugins	229.595s

before: ok  	k8s.io/kubernetes/test/integration/scheduler/plugins	320.507s

File: test/integration/util/util.go
Comment (kerthcet) [reply test/integration/util/util.go:201]: Seems only used by scheduler. Let me fix this in another PR to make 
- construct scheduler
- start informer
- start scheduler

a util function, this is widely used by lots of tests but we have to call these three functions one by one. etc. https://github.com/kubernetes/kubernetes/blob/742316ee2136b24b9edb0bb0d5bf69a6bad6f364/test/integration/volumescheduling/volume_capacity_priority_test.go#L50-L52

File: test/integration/util/util.go
Comment (kerthcet) [reply test/integration/util/util.go:214]: It's right, but not straightforward. The context is the top one in initializing the test environment so it will definitely kill the scheduler. Let me polish this comment.

File: test/integration/util/util.go
Comment (kerthcet) [reply test/integration/util/util.go:156]: CloseFn mostly cares about the apiserver, calling the CancelFn will not close the apiserver. See here https://github.com/kubernetes/kubernetes/blob/742316ee2136b24b9edb0bb0d5bf69a6bad6f364/test/integration/framework/test_server.go#L68-L86


The context to construct a apiserver is derived from the root context, I was meant to pass a context to `StartTestServer`, but that touches a lot of tests.

File: test/integration/scheduler/plugins/plugins_test.go
Comment (kerthcet) [reply test/integration/scheduler/plugins/plugins_test.go:2505]: Rename to `cleanupSchedulerFrameworkTest` vs `initTestSchedulerForFrameworkTest`.

File: test/integration/scheduler/plugins/plugins_test.go
Comment (pohly) [inline test/integration/scheduler/plugins/plugins_test.go:2475]: This mentions nodes and pods. Why just those? What about deleting namespaces?

The comment explains the "what" (which can be seen from the code), but not the less obvious "why". Can you add which objects need to be deleted here, now and in the future?

Comment (pohly) [reply test/integration/scheduler/plugins/plugins_test.go:2475]: Instead of having some code elsewhere create something and adding separate cleanup code for it here, can we make it so that the code which creates something also has the corresponding cleanup code and registers a callback for it?

This would be conceptually the same as `gingko.DeferCleanup`.

File: test/integration/util/util.go
Comment (pohly) [reply test/integration/util/util.go:156]: While investigating slow apiserver shutdown, I stumbled over `StartTestServer` not taking a context. That makes it impossible to set a runtime limit (test cancelation) and prevents per-test log output.

`StartTestServer` really should take a context. We shouldn't let existing code hold us back, that's how technical debt accumulates.

Can you add the context in your PR?

File: test/integration/scheduler/plugins/plugins_test.go
Comment (kerthcet) [reply test/integration/scheduler/plugins/plugins_test.go]: Update the comment and make the teardown func an output parameter. As you expected?

File: test/integration/util/util.go
Comment (kerthcet) [reply test/integration/util/util.go:156]: sgtm.

Comment (kerthcet) [reply test/integration/util/util.go]: Updated,  let's see whether the CI is green.

File: test/integration/scheduler/plugins/plugins_test.go
Comment (pohly) [inline test/integration/scheduler/plugins/plugins_test.go:2454]: Can you add a comment explaining under which circumstances the shutdown function should get called and when?

Some tests seem to ignore it (e.g. `TestActivatePods`).

File: test/integration/scheduler/plugins/plugins_test.go
Comment (pohly) [inline test/integration/scheduler/plugins/plugins_test.go:2067]: Do all tests always call `testutils.CleanupTest`?

If yes, then the code it executed might fit into a `t.Cleanup` callback that gets registered inside `testutils.InitTestAPIServer`.

File: test/integration/apiserver/apiserver_test.go
Comment (pohly) [reply test/integration/apiserver/apiserver_test.go:150]: Beware, I wasn't precise: NewTestContext does not handle cancellation. That still needs to be handled separately.

**Comment by alculquicondor** [comment]
FYI: we are past code freeze, so this can only merge in 1.28

**Comment by alculquicondor** [comment]
/hold
in case the command above fails on me

File: test/integration/apiserver/apiserver_test.go
Comment (kerthcet) [reply test/integration/apiserver/apiserver_test.go:150]: I would always like to use a withCancel context to make sure derived context will be cancelled in the final. Then it should be ?

```
_, ctx := ktesting.NewTestContext(t)
ctx, cancel := context.WithContext(ctx)
```

Comment (kerthcet) [reply test/integration/apiserver/apiserver_test.go:150]: This will touch a lot of codes, let's make it explicit before the change. By the way, `ktesting.NewTestContext(t)` seems used for tests, could you explain what can we benefit from this?

Comment (kerthcet) [reply test/integration/apiserver/apiserver_test.go:150]: Instead of create a context for each test, I'll return a context from setup. Easy to evolve in the future.

File: test/integration/scheduler/plugins/plugins_test.go
Comment (kerthcet) [reply test/integration/scheduler/plugins/plugins_test.go]: Moved inside the `InitTestAPIServer`. PTAL.

**Comment by kerthcet** [comment]
Addressed the comments, PTAL @pohly

**Review by alculquicondor — COMMENTED** [review]
oh, I forgot to send my comment :)

File: test/integration/scheduler/plugins/plugins_test.go
Comment (pohly) [inline test/integration/scheduler/plugins/plugins_test.go:2466]: This will result on the infamous "timed out waiting for the condition" error when something goes wrong.

When waiting for some condition, Gomega is better. Its failure message will include information about the latest error (if there was any while checking the condition) and the last negative condition check result.

https://github.com/kubernetes/kubernetes/pull/116207/commits/fc8198c08ffce91aa95f2a2fb30d7d352feac79f shows how to use `gomega.NewGomegaWithT(b).Eventually(...).Should(gomega.BeEmpty())` to check that pods, namespaces and claims got removed.

Integration tests are very similar to E2E tests in this regard, the only difference is that they are written with `go test` instead of Ginkgo.

I wonder whether it would be possible to consolidate... but probably not now. :sweat_smile:

Comment (kerthcet) [reply test/integration/scheduler/plugins/plugins_test.go:2466]: Let me track this with another issue, I think this is not an isolated case, we'd better to cover them once and for all.
And it's a big one already...

File: test/integration/util/util.go
Comment (kerthcet) [reply test/integration/util/util.go:200]: Updated.

**Comment by kerthcet** [comment]
/retest

Test failed by https://github.com/kubernetes/kubernetes/issues/107414

**Comment by kerthcet** [comment]
@pohly Plz take a look when you have time, thanks.

File: test/integration/scheduler/plugins/plugins_test.go
Comment (pohly) [reply test/integration/scheduler/plugins/plugins_test.go:2433]: How was this resolved?

File: test/integration/scheduler/plugins/plugins_test.go
Comment (pohly) [inline test/integration/scheduler/plugins/plugins_test.go:2452]: This seems to be the explanation how the `ShutdownFunc` that gets returned by `initTestSchedulerForFrameworkTest` should be used.

That explanation should be in the function documentation of `initTestSchedulerForFrameworkTest`, not in the middle of its implementation.

File: test/integration/scheduler/plugins/plugins_test.go
Comment (pohly) [reply test/integration/scheduler/plugins/plugins_test.go]: This is the shutdown function returned by `initTestSchedulerForFrameworkTest`, right? That looks like a better approach than the stand-alone `cleanupTestSchedulerForFrameworkTest` function here.

**Review by pohly — CHANGES_REQUESTED** [review]
Thinking aloud here... should all of these commits have been squashed into one?

I went through them one-by-one because I expected a large PR like this to be broken down into smaller, self-contained commit. But then I realized that I was re-reviewing old commits that got fixed in later commits, so I had to do a lot of back-and-forth to check whether a comment that I was about to make still applied.

But that's just a side note. Overall the change looks good, just one nit about documentation.

**Comment by kerthcet** [comment]
> Thinking aloud here... should all of these commits have been squashed into one?
> 
> I went through them one-by-one because I expected a large PR like this to be broken down into smaller, self-contained commit. But then I realized that I was re-reviewing old commits that got fixed in later commits, so I had to do a lot of back-and-forth to check whether a comment that I was about to make still applied.
> 
> But that's just a side note. Overall the change looks good, just one nit about documentation.

em.. what I learned from the community is not squash for better reviewing, or you can't know what changes later ... And you can organize the commits in the final. Do you think this not works for you?

**Comment by kerthcet** [comment]
I got the point, because you review the PR from the scratch, but I thought you have the reviewing context. next time I can organize the commits at first if you want.

**Comment by pohly** [comment]
I tend to prefer clean commits (can be merged right away when they are ready, can start from scratch when I need to, for example after coming back to a PR after a while) and use the GitHub features (changes since last review, force-push diff) to track what gets changed overall. Perhaps that's just me :shrug:

**Comment by alculquicondor** [comment]
Maybe you can split this PR in 2: one for the framework changes (plus one test update) and one for the rest of the test updates.

I think the way you structured the commits might allow to do this somewhat easily.

**Comment by alculquicondor** [comment]
If you are ok with reviewing as-is, we can squash. Otherwise, better split in 2 PRs as described above.

**Comment by pohly** [comment]
I'm fine with squashing into reasonable commits. My only remaining comment is about documenting `initTestSchedulerForFrameworkTest` which will be easy to check.

**Comment by kerthcet** [comment]
> This is the shutdown function returned by initTestSchedulerForFrameworkTest, right? That looks like a better approach than the stand-alone cleanupTestSchedulerForFrameworkTest function here.

I can't reply on the comment https://github.com/kubernetes/kubernetes/pull/116634#discussion_r1152971097, but the function is already removed, so let it go.

File: test/integration/scheduler/plugins/plugins_test.go
Comment (kerthcet) [reply test/integration/scheduler/plugins/plugins_test.go:2452]: Moved to the head.

**Comment by kerthcet** [comment]
Let me split into 2 PRs, 1) for the scheduler integration tests, 2) for the rest test files about passing a context to the StartTestServer. 1) is what I planned at first, 2) is something we agreed later, that can be a separate PR. This is more clear to me also.

File: test/integration/util/util.go
Comment (pohly) [inline test/integration/util/util.go:354]: I think a `ktesting.NewTestContext` is missing here.

Comment (kerthcet) [reply test/integration/util/util.go:354]: Yes, thanks for catching this. updated.

**Comment by pohly** [comment]
The first three commits are not clean yet: the first one mixes functional changes with dead code removal, and the last one fixes something in an earlier commit. I think you can simply squash all of them into one. The last commit is self-contained.

**Review by pohly — COMMENTED** [review]
I think all of these findings from strict linting should get fixed:
```
 test/integration/util/util.go:214:53: Error return value of `(k8s.io/client-go/kubernetes/typed/core/v1.NodeInterface).DeleteCollection` is not checked (errcheck)
	testCtx.ClientSet.CoreV1().Nodes().DeleteCollection(testCtx.Ctx, *metav1.NewDeleteOptions(0), metav1.ListOptions{})
	                                                   ^
test/integration/apiserver/export_test.go:39:41: Error return value of `(k8s.io/client-go/kubernetes/typed/core/v1.NamespaceInterface).Delete` is not checked (errcheck)
		clientSet.CoreV1().Namespaces().Delete(ctx, "export-fail", metav1.DeleteOptions{})
		                                      ^
test/integration/apiserver/flowcontrol/concurrency_test.go:52:55: captLocal: `MaxMutatingRequestsInFlight' should not be capitalized (gocritic)
func setup(t testing.TB, maxReadonlyRequestsInFlight, MaxMutatingRequestsInFlight int) (context.Context, *rest.Config, framework.TearDownFunc) {
                                                      ^
test/integration/scheduler/plugins/plugins_test.go:2464:13: SA1019: wait.Poll is deprecated: This method does not return errors from context, use PollWithContextTimeout. Note that the new method will no longer return ErrWaitTimeout and instead return errors defined by the context package. Will be removed in a future release. (staticcheck)
		if err := wait.Poll(time.Millisecond, wait.ForeverTestTimeout,
		
```

**Comment by kerthcet** [comment]
The `pull-kubernetes-verify-strict-lint` requires a lot of code changes.

File: test/integration/scheduler/filters/filters_test.go
Comment (pohly) [inline test/integration/scheduler/filters/filters_test.go:1513]: Is it possible that `testCtx.Ctx` has been canceled when it is time to clean up?

Comment (kerthcet) [reply test/integration/scheduler/filters/filters_test.go:1513]: No, testCtx.Ctx will be cancelled in t.Cleanup(), after defer, and it's first added, last called.

File: test/integration/scheduler/plugins/plugins_test.go
Comment (sathyanarays) [inline test/integration/scheduler/plugins/plugins_test.go:699]: nit: We are not creating API server everytime, right?

Comment (kerthcet) [reply test/integration/scheduler/plugins/plugins_test.go:699]: Yes, updated and applied to all. Thanks.

**Comment by kerthcet** [comment]
Resolved the conflicts and needs another lgtm.

**Comment by kerthcet** [comment]
Would you mind to label LGTM again @pohly

**Review by pohly — APPROVED** [review]
/lgtm
/approve

**Comment by kerthcet** [comment]
/assign @soltysh 
Would you mind to approve this as sig-apps approvers? Thanks.

**Comment by alculquicondor** [comment]
cc @kubernetes/sig-apps-leads 
It's important that we merge this PR soon to avoid blocking other PRs.

**Comment by kerthcet** [comment]
Kindly ping @kubernetes/sig-apps-leads

**Review by janetkuo — APPROVED** [review]

**Comment by alculquicondor** [comment]
maybe needs rebase?

**Review by pohly — APPROVED** [review]
/lgtm

**Pull request merged**
